rollbar_dart/lib/src/rollbar_infrastructure.dart

```diff
@@ -1,122 +1,122 @@
-    _sendPort.send(rollbarConfig);
-    return _sendPort;
+  static Future<RollbarInfrastructure> start({required Config config}) async {
+    final receivePort = ReceivePort();
+    final isolate = await Isolate.spawn(work, receivePort.sendPort);
+    final sendPort = await receivePort.first;
+    // [todo] figure a cleanear way of removing the sender/transformer
+    // before, isolates imposes limitations where the transformer/senders
+    // functions must be free functions, but the latest architecture changes
+    // allow us to init everything that requires these functions before the
+    // isolates, so there's no more need for the Config at this point to carry
+    // these functions, thus eliminating the restriction/limitations.
+    sendPort.send(Config.fromMap(config.toMap()));
+    return RollbarInfrastructure._(isolate, receivePort, sendPort);
   }
 
   Future<void> dispose() async {
     // Send a signal to the spawned isolate indicating that it should exit:
     _sendPort.send(null);
     _receivePort.close();
+    _isolate.kill(priority: Isolate.beforeNextEvent);
   }
 
-  static final RollbarInfrastructure instance = RollbarInfrastructure._();
+  //static final RollbarInfrastructure instance = RollbarInfrastructure._();
 
   void process({required PayloadRecord record}) {
     _sendPort.send(record);
   }
 
-  static Future<void> _processWorkItemsInBackground(SendPort sendPort) async {
-    ModuleLogger.moduleLogger.info('Infrastructure isolate started.');
-
-    // Send a SendPort to the main isolate (RollbarInfrastructure)
-    // so that it can send JSON strings to this isolate:
-    final receivePort = ReceivePort();
-    sendPort.send(receivePort.sendPort);
+  @internal
+  static Future<void> work(SendPort sendPort) async {
+    final infrastructurePort = ReceivePort();
+    sendPort.send(infrastructurePort.sendPort);
 
     // Wait for messages from the main isolate.
-    await for (final message in receivePort) {
+    await for (final message in infrastructurePort) {
       bool continueProcessing = await _process(message);
       if (!continueProcessing) {
         break;
       }
     }
-
-    ModuleLogger.moduleLogger.info('Infrastructure isolate finished.');
-    Isolate.exit();
   }
 
   static Future<bool> _process(dynamic message) async {
-    // [HACK] This delay helps to avoid occasional sqlite's "db locked"
-    // errors/exceptions. Keep it until we figure out why sqlite has these
-    // errors randomly:
-    await Future.delayed(Duration(milliseconds: 25));
-
-    if (message is Config) {
-      _processConfig(message);
-      await _processAllPendingRecords();
-      return true;
-    } else if (message is PayloadRecord) {
-      await _processPayloadRecord(message);
-      return true;
-    } else if (message == null) {
-      // Exit if the main isolate sends a null message, indicating
-      // it is the time to exit.
+    if (message == null) {
```

Comment:
Unfortunately, I can't pinpoint an exact piece of code or mechanism that made it fail. I made multiple changes to the way we approach multi-threading in our library which seem to have fixed this issue for good:

The Dart SQLite library has various limitations due to Dart not having a true concept of "multithreading." 

The main limitation is that it can't handle concurrents read and write and errors out with the infamous "db locked" error when trying to do one while the other is still on it.
Source: https://pub.dev/documentation/sqflite/latest/
> Due to the way transaction works in SQLite (threads), concurrent read and write transaction are not supported.

Another issue is that it is recommended that the amount of Dart `Isolate`s not surpass the amount of CPU cores. This is because Dart runs each Isolate on a separate CPU core. iPhones have 2 high-frequency cores and 4 high-efficiency cores, while Android varies from 2 to 6.

We had running two `Isolate`s, the `Infrastructure` `Isolate` and the `UncaughtErrorHandler` Isolate, both were accessing the database. Now the latter is gone, so this _may_ have been the actual source of the problem.

The other potential issue is that the the mutable nature of how the data structures used to be, may mean that the payload the database received, _could have_ suffered some mutation we were not aware of along the way.